hbase-client/src/main/java/org/apache/hadoop/hbase/client/ZKConnectionRegistry.java

```diff
@@ -17,235 +17,35 @@
  */
 package org.apache.hadoop.hbase.client;
 
-import static org.apache.hadoop.hbase.client.RegionInfo.DEFAULT_REPLICA_ID;
-import static org.apache.hadoop.hbase.client.RegionInfoBuilder.FIRST_META_REGIONINFO;
-import static org.apache.hadoop.hbase.client.RegionReplicaUtil.getRegionInfoForDefaultReplica;
-import static org.apache.hadoop.hbase.client.RegionReplicaUtil.getRegionInfoForReplica;
-import static org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil.lengthOfPBMagic;
-import static org.apache.hadoop.hbase.trace.TraceUtil.tracedFuture;
-import static org.apache.hadoop.hbase.util.FutureUtils.addListener;
-import static org.apache.hadoop.hbase.zookeeper.ZKMetadata.removeMetaData;
-
-import java.io.IOException;
-import java.util.List;
-import java.util.concurrent.CompletableFuture;
-import java.util.stream.Collectors;
-import org.apache.commons.lang3.mutable.MutableInt;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.hbase.ClusterId;
-import org.apache.hadoop.hbase.HRegionLocation;
-import org.apache.hadoop.hbase.RegionLocations;
-import org.apache.hadoop.hbase.ServerName;
-import org.apache.hadoop.hbase.exceptions.DeserializationException;
-import org.apache.hadoop.hbase.master.RegionState;
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
 import org.apache.hadoop.hbase.security.User;
-import org.apache.hadoop.hbase.util.Pair;
-import org.apache.hadoop.hbase.zookeeper.ReadOnlyZKClient;
-import org.apache.hadoop.hbase.zookeeper.ZNodePaths;
 import org.apache.yetus.audience.InterfaceAudience;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import org.apache.hadoop.hbase.shaded.protobuf.generated.HBaseProtos;
-import org.apache.hadoop.hbase.shaded.protobuf.generated.ZooKeeperProtos;
-
 /**
  * Zookeeper based registry implementation.
+ * @deprecated As of 2.6.0, replaced by {@link RpcConnectionRegistry}, which is the default connection mechanism as of 3.0.0. Expected to be removed in 4.0.0.
+ * @see <a href="https://issues.apache.org/jira/browse/HBASE-23324">HBASE-23324</a> and its parent ticket for details.
  */
-@InterfaceAudience.Private
-class ZKConnectionRegistry implements ConnectionRegistry {
-
+@Deprecated
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.CONFIG)
+class ZKConnectionRegistry extends ZKConnectionRegistryInternal {
```

Comment:
The diff isn't rendering well, but I renamed the existing `ZKConnectionRegistry` to `ZKConnectionRegistryInternal` and then created the new class named `ZKConnectionRegistry` that subclasses and adds the warning.

Reply:
The idea is that we can remove this class, but keep the internal on if we want to in 4.0, right ?

Reply:
Correct. HMaster and RegionServers will continue to use the ZKConnectionRegistry. Once the public version is removed, we can do as we like with the internal version, such as move it to hbase-server.jar.

Reply:
I think in 3.0.0, HMaster and HRegionServer also uses RpcConnectionRegistry?

Reply:
> I think in 3.0.0, HMaster and HRegionServer also uses RpcConnectionRegistry?

Having read through the details of `ShortCircuitConnectionRegistry`, I now believe that you are correct. Nice!

So then this rename and Internal stuff is irrelevant -- we can drop the ZKConnectionRegistry entirely. WDYT?

Reply:
For me I prefer we remove it entirely, in 4.0.0.